servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/NewRoundRobinLoadBalancer.java

```diff
@@ -250,59 +251,88 @@ public void onNext(@Nullable final Collection<? extends ServiceDiscovererEvent<R
+                sendReadyEvent = false;
+
+                // First we make a map of addresses to events so that we don't get quadratic behavior for diffing.
+                // Unfortunately we need to make this every iteration of the CAS loop since we remove entries
+                // for hosts that already exist. If this results in to many collisions and map rebuilds we should
+                // re-assess how we manage concurrency for list mutations.
+                final Map<ResolvedAddress, ServiceDiscovererEvent<ResolvedAddress>> eventMap = new HashMap<>();
+                for (ServiceDiscovererEvent<ResolvedAddress> event : events) {
+                    ServiceDiscovererEvent<ResolvedAddress> old = eventMap.put(event.address(), event);
+                    if (old != null) {
+                        LOGGER.error("Multiple ServiceDiscoveryEvent's detected for address {}. Event: {}.",
+                                event.address(), event);
+                    }
+                }
 
-                if (AVAILABLE.equals(eventStatus)) {
-                    if (usedAddresses.size() == 1) {
-                        eventStreamProcessor.onNext(LOAD_BALANCER_READY_EVENT);
+                // First thing we do is go through the existing hosts and see if we need to transfer them. These
+                // will be all existing hosts that either don't have a matching discovery event or are not marked
+                // as unavailable. If they are marked unavailable, we need to close them (which is idempotent).
+                for (Host<ResolvedAddress, C> host : usedHosts) {
+                    ServiceDiscovererEvent<ResolvedAddress> event = eventMap.remove(host.address());
+                    if (event == null) {
+                        // Host doesn't have a SD update so just copy it over.
+                        nextHosts.add(host);
+                    } else if (AVAILABLE.equals(event.status())) {
+                        sendReadyEvent = true;
+                        // If the host is already in CLOSED state, we should discard it and create a new entry.
+                        // For duplicate ACTIVE events or for repeated activation due to failed CAS
+                        // of replacing the usedHosts array the marking succeeds so we will not add a new entry.
+                        if (host.markActiveIfNotClosed()) {
+                            nextHosts.add(host);
+                        } else {
+                            nextHosts.add(createHost(event.address()));
+                        }
+                    } else if (EXPIRED.equals(event.status())) {
+                        if (!host.markExpired()) {
+                            nextHosts.add(host);
+                        }
+                    } else if (UNAVAILABLE.equals(event.status())) {
+                        host.markClosed();
+                    } else {
+                        LOGGER.error("{}: Unexpected Status in event:" +
```

Comment:
1. Same here, users can be paranoid about `error` logs. Consider `warn` or `info` for this case.
2. Instead of `Unexpected`, consider using `Unsupported`. Bcz SD and LB are independent, technically they can use statuses that one or another does not know about.